### ISSUE TITLE
Support for developers using Eclipse

### ISSUE THREAD

**Issue by dsyer** [ratpack/ratpack#658]
Something changed in #446, but it's still not possible to import ratpack directly into a stock Spring Tool Suite. Several problems seem to be plaguing me, compiler errors in Java can mostly be fixed with simple re-arrangements of the code (maybe a PR for that?). I also had to manually edit the `.classpath` for `ratpack-test-internal` to remove the "src" folder from the build path (no idea what that was doing there). The last problem which I can't see a solution for at the minute is a load of cryptic errors from the groovy compiler.

**Comment by dsyer** [comment]
Also getting these on startup

```
java.lang.StackOverflowError
	at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:315)
	at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:339)
	at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:339)
	at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:339)
	at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:339)
	at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:339)
	at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:339)
	at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:339)
	at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:339)
	at org.codehaus.groovy.ast.tools.GenericsUtils.correctToGenericsSpecRecurse(GenericsUtils.java:339)
...
```

**Comment by dsyer** [comment]
The Groovy compiler/build barfs like this on ratpack-core:

```
General error during semantic analysis: nyi class org.eclipse.jdt.internal.compiler.lookup.WildcardBinding

java.lang.IllegalStateException: nyi class org.eclipse.jdt.internal.compiler.lookup.WildcardBinding
	at org.codehaus.jdt.groovy.internal.compiler.ast.JDTClassNodeBuilder.toRawType(JDTClassNodeBuilder.java:163)
	at org.codehaus.jdt.groovy.internal.compiler.ast.JDTClassNode.makeParameter(JDTClassNode.java:392)
	at org.codehaus.jdt.groovy.internal.compiler.ast.JDTClassNode.makeParameters(JDTClassNode.java:381)
	at org.codehaus.jdt.groovy.internal.compiler.ast.JDTClassNode.methodBindingToMethodNode(JDTClassNode.java:354)
	at org.codehaus.jdt.groovy.internal.compiler.ast.JDTClassNode.initializeMembers(JDTClassNode.java:307)
	at org.codehaus.jdt.groovy.internal.compiler.ast.JDTClassNode.initialize(JDTClassNode.java:227)
	at org.codehaus.jdt.groovy.internal.compiler.ast.JDTClassNode.lazyClassInit(JDTClassNode.java:158)
	at org.codehaus.groovy.ast.ClassNode.getDeclaredMethods(ClassNode.java:971)
	at org.codehaus.groovy.ast.ClassNode.getMethods(ClassNode.java:986)
	at org.codehaus.groovy.ast.ClassNode.hasPossibleStaticMethod(ClassNode.java:1417)
	at org.codehaus.groovy.control.StaticImportVisitor.findStaticMethod(StaticImportVisitor.java:549)
	at org.codehaus.groovy.control.StaticImportVisitor.findStaticMethodImportFromModule(StaticImportVisitor.java:437)
	at org.codehaus.groovy.control.StaticImportVisitor.transformMethodCallExpression(StaticImportVisitor.java:245)
	at org.codehaus.groovy.control.StaticImportVisitor.transform(StaticImportVisitor.java:83)
	at org.codehaus.groovy.ast.expr.Expression.transformExpressions(Expression.java:49)
	at org.codehaus.groovy.ast.expr.ArgumentListExpression.transformExpression(ArgumentListExpression.java:67)
	at org.codehaus.groovy.control.StaticImportVisitor.transform(StaticImportVisitor.java:92)
	at org.codehaus.groovy.control.StaticImportVisitor.transformMethodCallExpression(StaticImportVisitor.java:232)
	at org.codehaus.groovy.control.StaticImportVisitor.transform(StaticImportVisitor.java:83)
	at org.codehaus.groovy.ast.ClassCodeExpressionTransformer.visitExpressionStatement(ClassCodeExpressionTransformer.java:139)
	at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:40)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitBlockStatement(CodeVisitorSupport.java:37)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitBlockStatement(ClassCodeVisitorSupport.java:206)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:69)
	at org.codehaus.groovy.control.StaticImportVisitor.transformClosureExpression(StaticImportVisitor.java:321)
	at org.codehaus.groovy.control.StaticImportVisitor.transform(StaticImportVisitor.java:86)
	at org.codehaus.groovy.ast.expr.Expression.transformExpressions(Expression.java:49)
	at org.codehaus.groovy.ast.expr.ArgumentListExpression.transformExpression(ArgumentListExpression.java:67)
	at org.codehaus.groovy.control.StaticImportVisitor.transform(StaticImportVisitor.java:92)
	at org.codehaus.groovy.control.StaticImportVisitor.transformMethodCallExpression(StaticImportVisitor.java:232)
	at org.codehaus.groovy.control.StaticImportVisitor.transform(StaticImportVisitor.java:83)
	at org.codehaus.groovy.ast.ClassCodeExpressionTransformer.visitExpressionStatement(ClassCodeExpressionTransformer.java:139)
	at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:40)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitBlockStatement(CodeVisitorSupport.java:37)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitBlockStatement(ClassCodeVisitorSupport.java:206)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:69)
	at org.codehaus.groovy.control.StaticImportVisitor.transformClosureExpression(StaticImportVisitor.java:321)
	at org.codehaus.groovy.control.StaticImportVisitor.transform(StaticImportVisitor.java:86)
	at org.codehaus.groovy.ast.expr.Expression.transformExpressions(Expression.java:49)
	at org.codehaus.groovy.ast.expr.ArgumentListExpression.transformExpression(ArgumentListExpression.java:67)
	at org.codehaus.groovy.control.StaticImportVisitor.transform(StaticImportVisitor.java:92)
	at org.codehaus.groovy.control.StaticImportVisitor.transformMethodCallExpression(StaticImportVisitor.java:232)
	at org.codehaus.groovy.control.StaticImportVisitor.transform(StaticImportVisitor.java:83)
	at org.codehaus.groovy.ast.ClassCodeExpressionTransformer.visitExpressionStatement(ClassCodeExpressionTransformer.java:139)
	at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:40)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitBlockStatement(CodeVisitorSupport.java:37)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitBlockStatement(ClassCodeVisitorSupport.java:206)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:69)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitClassCodeContainer(ClassCodeVisitorSupport.java:103)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitConstructorOrMethod(ClassCodeVisitorSupport.java:114)
	at org.codehaus.groovy.ast.ClassCodeExpressionTransformer.visitConstructorOrMethod(ClassCodeExpressionTransformer.java:50)
	at org.codehaus.groovy.control.StaticImportVisitor.visitConstructorOrMethod(StaticImportVisitor.java:59)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitMethod(ClassCodeVisitorSupport.java:125)
	at org.codehaus.groovy.ast.ClassNode.visitContents(ClassNode.java:1200)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitClass(ClassCodeVisitorSupport.java:52)
	at org.codehaus.groovy.control.StaticImportVisitor.visitClass(StaticImportVisitor.java:53)
	at org.codehaus.groovy.control.CompilationUnit$2.call(CompilationUnit.java:722)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1216)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:646)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:624)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:601)
	at org.codehaus.jdt.groovy.internal.compiler.ast.GroovyCompilationUnitDeclaration.processToPhase(GroovyCompilationUnitDeclaration.java:202)
	at org.codehaus.jdt.groovy.internal.compiler.ast.GroovyCompilationUnitDeclaration.resolve(GroovyCompilationUnitDeclaration.java:2202)
	at org.eclipse.jdt.internal.compiler.Compiler.process(Compiler.java:827)
	at org.eclipse.jdt.internal.compiler.ProcessTaskManager.run(ProcessTaskManager.java:137)
	at java.lang.Thread.run(Thread.java:744)
```

**Comment by dsyer** [comment]
And this one in `ratpack-groovy` (`TestCompileStaticApp.groovy`):

```
Groovy:[Static type checking] - Cannot find matching method ratpack.groovy.TestCompileStaticApp#get(groovy.lang.Closure). Please check if the declared type is right and if the method exists.
```

**Comment by dsyer** [comment]
Removing `IoUtilSpec.groovy` from `ratpack-core` allows that project to build correctly. The `ratpack-groovy` problem seems to be a show stopper because it's a utility used all over the tests in other modules. Is it a Java 8 method literal being used as a Groovy closure or something? I can't even parse the code by eye.

**Comment by dsyer** [comment]
Another (unrelated?) problem: you can't run any tests in Eclipse:

```
org.spockframework.runtime.InvalidSpecException: Specification 'ratpack.test.internal.EmbeddedRatpackSpec' was not compiled properly (Spock AST transform was not run); try to do a clean build
	at org.spockframework.runtime.SpecUtil.checkIsSpec(SpecUtil.java:53)
	at org.spockframework.runtime.SpecInfoBuilder.buildSpec(SpecInfoBuilder.java:74)
	at org.spockframework.runtime.SpecInfoBuilder.doBuild(SpecInfoBuilder.java:56)
	at org.spockframework.runtime.SpecInfoBuilder.buildSuperSpec(SpecInfoBuilder.java:68)
	at org.spockframework.runtime.SpecInfoBuilder.doBuild(SpecInfoBuilder.java:55)
	at org.spockframework.runtime.SpecInfoBuilder.buildSuperSpec(SpecInfoBuilder.java:68)
	at org.spockframework.runtime.SpecInfoBuilder.doBuild(SpecInfoBuilder.java:55)
	at org.spockframework.runtime.SpecInfoBuilder.buildSuperSpec(SpecInfoBuilder.java:68)
	at org.spockframework.runtime.SpecInfoBuilder.doBuild(SpecInfoBuilder.java:55)
	at org.spockframework.runtime.SpecInfoBuilder.build(SpecInfoBuilder.java:42)
	at org.spockframework.runtime.Sputnik.getSpec(Sputnik.java:78)
	at org.spockframework.runtime.Sputnik.runExtensionsIfNecessary(Sputnik.java:86)
	at org.spockframework.runtime.Sputnik.getDescription(Sputnik.java:55)
	at org.junit.runner.Runner.testCount(Runner.java:38)
...
```

**Comment by johnrengelman** [comment]
It appears that most of the errors are associate with various type conversions not being accepted by `ecj`.

https://bugs.eclipse.org/bugs/show_bug.cgi?id=434118
https://bugs.eclipse.org/bugs/show_bug.cgi?id=434483

etc.

I tried running the latest milestone release of 4.5 and that fixes some issues but not all. Unfortunately, we can't do much about Eclipse having bugs around type inference and Lambda syntax.

**Comment by alkemist** [comment]
> I tried running the latest milestone release of 4.5 and that fixes some issues but not all. Unfortunately, we can't do much about Eclipse having bugs around type inference and Lambda syntax.

Well we can. We can change the code to be something palatable to `ecj`, but I don't want to do this without a good reason. As far as I am concerned, if it compiles with `javac` it's valid and I'm not interested in making it work with `ecj` just because.

However, @dsyer if this is blocking you from contributing some code that would be valuable enough to the project to justify us spending our personal time (i.e. Ratpack is all volunteer based) working through these Eclipse so you can contribute then we can spend some of the little time we have on this. Can you elaborate on what you are planning to contribute?

**Comment by dsyer** [comment]
The `ecj` issues are all trivial so far (just a few lines of additional casts and stuff like that, so the compiler can infer types more easily). I'll send a PR for what I have if you like. So far I only covered the core modules.

But the bigger problem really is the groovy compiler (especially given that no-one is supporting it in eclipse these days). I don't have any suggestions there. Maybe one of the groovy team could take a look?

The features I want to contribute are in spring-boot-ratpack (project in my github id). Again it's not huge, but it makes Ratpack optionally embeddable in a Spring Boot application and provides some autoconfiguration for common renderers (those feature can easily be split - the autoconfig might fit better in Spring Boot itself). I'm having some trouble upgrading to 0.9.15, so it'd be good to get some more eyes in it if anyone has some time.

**Comment by rhart** [comment]
@dsyer PR sent for upgrading to 0.9.15.  Hope it helps.

**Comment by alkemist** [comment]
@dsyer it would be really great to get that boot stuff in.

I'd be happy to merge a PR for the necessary Java issues.

I'm going to look into the Groovy issues and see what we can do. Even if they do get fixed in Groovy-Eclipse, I'd rather not wait.

Thanks for your offer to help out more, we'll definitely take you up on that :)

**Comment by dsyer** [comment]
OK, great. I did some work on it yesterday in my own repo, so it looks like the 0.9.15 upgrade is ready (thanks largely to @rhart). I can send a quick PR for the ecj issues I worked around, and then another for the Boot embedding features (and possibly a further one for the autoconfiguration). I don't have "ratpack standard" unit tests yet (just Spring Boot integration tests), so you might have to guide me a bit on that once the PR is done.
